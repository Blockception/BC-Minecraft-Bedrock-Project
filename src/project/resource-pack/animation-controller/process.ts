import { Types } from "bc-minecraft-bedrock-types";
import * as Internal from "../../../internal/resource-pack/animation-controller";
import { Documentation, TextDocument } from "../../../types";
import { References, Using } from "../../../types/references";
import { harvestMolang } from "../../molang";
import { AnimationController } from "./animation-controller";
import { Effect } from "../../../internal/resource-pack";

/** */
export function Process(doc: TextDocument): AnimationController[] | undefined {
  const imp = TextDocument.toObject(doc, Internal.AnimationControllers.is);
  if (!imp) return undefined;

  const uri = doc.uri;
  const content = doc.getText();
  const container = imp.animation_controllers;

  return Object.entries(container)
    .filter(([, controller]) => Internal.AnimationController.is(controller))
    .map(([id, controller]) => {
      const item: AnimationController = {
        id: id,
        location: Types.Location.create(uri, content.indexOf(id)),
        molang: harvestMolang(content, controller),
        documentation: Documentation.getDoc(doc, () => `RP Animation Controller: '${id}'`),
        animations: References.create(),
        particles: References.create(),
        sounds: References.create(),
      };

      Object.values(controller.states).map((state) => {
        Types.Conditional.forEach(state.animations, (reference) => item.animations.using.add(reference));

        if (state.particle_effects) harvest(state.particle_effects, item.particles);
        if (state.sound_effects) harvest(state.sound_effects, item.sounds);
      });

      return item;
    })
    .filter((item) => item !== undefined);
}

function harvest(data: Effect[], receiver: Using) {
  Using.add(
    receiver,
    data.map((e) => e.effect).filter((e) => e !== undefined)
  );
}
