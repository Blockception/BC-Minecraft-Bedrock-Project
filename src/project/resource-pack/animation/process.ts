import { Types } from "bc-minecraft-bedrock-types";
import * as Internal from "../../../internal/resource-pack";
import { Documentation, SMap, TextDocument } from "../../../types";
import { References } from "../../../types/references";
import { harvestMolang } from "../../molang/harvest";
import { Animation } from "./animation";

/** */
export function Process(doc: TextDocument): Animation[] | undefined {
  const imp = TextDocument.toObject(doc, Internal.Animations.is);
  if (!imp) return undefined;

  const uri = doc.uri;
  const content = doc.getText();
  const out: Animation[] = [];
  const container = imp.animations;
  const keys = Object.getOwnPropertyNames(container);

  for (let I = 0; I < keys.length; I++) {
    const id = keys[I];
    const anim = container[id];

    if (Internal.Animation.is(anim)) {
      const item: Animation = {
        id: id,
        location: Types.Location.create(uri, content.indexOf(id)),
        molang: harvestMolang(content, anim),
        documentation: Documentation.getDoc(
          doc,
          () => `RP Animation: '${id}', loop: ${anim.loop ?? false}, length: ${anim.animation_length ?? "unknown"}`
        ),
        particles: References.empty(),
        sounds: References.empty(),
      };

      if (anim.particle_effects)
        SMap.forEach(anim.particle_effects, (value) => processEffect(value, item.particles.using));
      if (anim.sound_effects) SMap.forEach(anim.sound_effects, (value) => processEffect(value, item.sounds.using));

      out.push(item);
    }
  }

  return out;
}

type effect_carrier = { effect?: string };

function processEffect(item: effect_carrier | effect_carrier[], receiver: string[]): void {
  const process = (item: effect_carrier) => {
    if (item.effect) receiver.push(item.effect);
  };

  if (Array.isArray(item)) {
    item.forEach(process);
  } else {
    process(item);
  }
}
