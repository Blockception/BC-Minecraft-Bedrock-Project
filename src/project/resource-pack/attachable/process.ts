import * as Internal from "../../../internal/resource-pack";
import { Json } from "../../../internal";
import { Molang } from "bc-minecraft-molang";
import { TextDocument } from "../../../types";
import { Attachable } from "./attachable";
import { DefinedUsing } from "bc-minecraft-molang";
import { Documentation } from "../../../types";
import { Types } from "bc-minecraft-bedrock-types";

/**
 *
 * @param doc
 * @returns
 */
export function Process(doc: TextDocument): Attachable | undefined {
  const uri = doc.uri;
  const content = doc.getText();
  const imp = Json.To<Internal.Attachable>(doc);

  if (!Internal.Attachable.is(imp)) return undefined;

  const container = imp["minecraft:attachable"];
  const description = container.description;
  const id = description.identifier;
  const out: Attachable = {
    id: id,
    location: Types.Location.create(uri, content.indexOf(id)),
    molang: Molang.MolangFullSet.harvest(container),
    animations: DefinedUsing.create(),
    documentation: Documentation.getDoc(doc, () => `Attachable Item: ${id}`),
  };
  Molang.MolangFullSet.fromScript(description.scripts ?? {}, out.molang);

  //process animations
  Types.Definition.forEach(description.animations, (reference, id) => {
    out.animations.defined.push(reference);
    out.animations.using.push(id);
  });

  //process Animation controller
  description.animation_controllers?.forEach((item) => {
    const temp = flatten(item);
    if (temp) out.animations.using.push(temp);
  });

  //process geometries
  Types.Definition.forEach(description.geometry, (reference, id) => {
    out.molang.geometries.defined.push(reference);
    out.molang.geometries.using.push(removePrefix(id));
  });

  //process materials
  Types.Definition.forEach(description.materials, (reference, id) => {
    out.molang.materials.defined.push(reference);
    out.molang.materials.using.push(removePrefix(id));
  });

  //process textures
  Types.Definition.forEach(description.textures, (reference, id) => {
    out.molang.textures.defined.push(reference);
    out.molang.textures.using.push(id);
  });

  return out;
}

function flatten(data: string | Types.Definition): string | undefined {
  if (typeof data === "string") return data;

  const key = Object.getOwnPropertyNames(data)[0];

  if (key) return data[key];

  return undefined;
}

function removePrefix(id: string): string {
  const index = id.indexOf(".");

  if (index > -1) return id.slice(index + 1);

  return id;
}