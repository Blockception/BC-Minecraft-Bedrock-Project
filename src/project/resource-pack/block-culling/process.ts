import { Types } from "bc-minecraft-bedrock-types";
import { Json } from "../../../internal";
import * as Internal from "../../../internal/resource-pack";
import { Defined, Documentation, TextDocument } from "../../../types";
import { BlockCulling } from "./culling";

export function Process(doc: TextDocument): BlockCulling | undefined {
  const uri = doc.uri;
  const content = doc.getText();
  const imp = Json.To<Internal.BlockCulling>(doc);
  if (!Internal.BlockCulling.is(imp)) return undefined;

  const id = imp["minecraft:block_culling_rules"].description.identifier;
  if (typeof id !== "string") return undefined;

  return {
    id: id,
    affected_bones: Defined.wrap(
      imp["minecraft:block_culling_rules"]?.rules?.map((r) => r?.geometry_part?.bone).filter((b) => b !== undefined)
    ),
    location: Types.Location.create(uri, content.indexOf(id)),
    documentation: Documentation.getDoc(doc, () => `Block culling rule: ${id}`),
  };
}
