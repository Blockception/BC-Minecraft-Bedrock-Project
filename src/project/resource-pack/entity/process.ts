import * as Internal from "../../../internal/resource-pack";
import { Json } from "../../../internal";
import { Molang } from "bc-minecraft-molang";
import { Types } from "bc-minecraft-bedrock-types";
import { TextDocument, Documentation } from "../../../types";
import { Entity } from "./entity";
import { DefinedUsing } from "bc-minecraft-molang";

/**
 *
 * @param doc
 * @returns
 */
export function Process(doc: TextDocument): Entity | undefined {
  const uri = doc.uri;
  const content = doc.getText();
  const imp = Json.To<Internal.Entity>(doc);

  if (!Internal.Entity.is(imp)) return undefined;

  const description = imp["minecraft:client_entity"].description;
  const id = description.identifier;
  const out: Entity = {
    id: id,
    location: Types.Location.create(uri, content.indexOf(id)),
    molang: Molang.MolangFullSet.harvest(description),
    animations: DefinedUsing.create(),
    documentation: Documentation.getDoc(doc, () => `Entity: ${id}`),
  };
  Molang.MolangFullSet.fromScript(description.scripts ?? {}, out.molang);

  //Process animations
  Types.Definition.forEach(description.animations, (reference, id) => {
    out.animations.defined.push(reference);
    out.animations.using.push(id);
  });

  //Process Animation controller
  description.animation_controllers?.forEach((item) => {
    const temp = flatten(item);
    if (temp) out.animations.using.push(temp);
  });

  //Process geometries
  Types.Definition.forEach(description.geometry, (reference, id) => {
    out.molang.geometries.defined.push(reference);
    out.molang.geometries.using.push(removePrefix(id));
  });

  //Process materials
  Types.Definition.forEach(description.materials, (reference, id) => {
    out.molang.materials.defined.push(reference);
    out.molang.materials.using.push(removePrefix(id));
  });

  //Process textures
  Types.Definition.forEach(description.textures, (reference, id) => {
    out.molang.textures.defined.push(reference);
    out.molang.textures.using.push(id);
  });

  return out;
}

function flatten(data: string | Types.Definition): string | undefined {
  if (typeof data === "string") return data;

  const key = Object.getOwnPropertyNames(data)[0];

  if (key) return data[key];

  return undefined;
}

function removePrefix(id: string): string {
  const index = id.indexOf(".");

  if (index > -1) return id.slice(index + 1);

  return id;
}
