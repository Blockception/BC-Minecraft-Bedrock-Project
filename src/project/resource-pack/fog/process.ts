import * as Internal from "../../../internal/resource-pack";
import { Types } from "bc-minecraft-bedrock-types";
import { TextDocument } from "../../../types";
import { Fog } from "./fog";
import { Documentation } from "../../../types";

/**
 *
 * @param doc
 * @returns
 */
export function Process(doc: TextDocument): Fog | undefined {
  const imp = TextDocument.toObject(doc, Internal.Fog.is);
  if (!imp) return undefined;

  const uri = doc.uri;
  const content = doc.getText();
  const container = imp["minecraft:fog_settings"];
  const id = container.description.identifier;

  return {
    id: id,
    location: Types.Location.create(uri, content.indexOf(id)),
    documentation: Documentation.getDoc(doc, () => `Fog: ${id}`),
  };
}
