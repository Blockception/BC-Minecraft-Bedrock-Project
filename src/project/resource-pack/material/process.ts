import { Types } from "bc-minecraft-bedrock-types";
import { Json } from "../../../internal/json";
import * as internal from "../../../internal/resource-pack/material";
import { Documentation, TextDocument } from "../../../types";
import { Material } from "./material";

/**
 *
 * @param doc
 * @returns
 */
export function process(doc: TextDocument): Material[] | undefined {
  const uri = doc.uri;
  const content = doc.getText();
  const imp = Json.To<internal.Material>(doc);

  if (!internal.Material.is(imp)) return undefined;

  return Object.entries(imp)
    .filter(([key]) => key !== "format_version")
    .map(([key]) => {
      return {
        id: key,
        location: Types.Location.create(uri, content.indexOf(key)),
        documentation: Documentation.getDoc(doc, () => `Material: ${key}`),
      };
    });
}
