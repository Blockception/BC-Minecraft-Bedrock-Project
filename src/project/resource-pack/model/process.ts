import * as internal from "../../../internal/resource-pack/model";
import { Json } from "../../../internal/json";
import { Types } from "bc-minecraft-bedrock-types";
import { TextDocument } from "../../../types";
import { Model } from "./model";
import { Documentation } from "../../../types";

/**
 *
 * @param doc
 * @returns
 */
export function Process(doc: TextDocument): Model[] | undefined {
  const uri = doc.uri;
  const content = doc.getText();
  const imp = Json.To<internal.Model>(doc);

  if (!internal.Model.is(imp)) return undefined;
  const out: Model[] = [];

  for (const [key, value] of Object.entries(imp)) {
    if (!key.startsWith("geometry.")) continue;
    if (!internal.ModelLegacySpec.is(value)) continue;

    out.push(
      createModel({
        id: key,
        location: Types.Location.create(uri, content.indexOf(key)),
        documentation: Documentation.getDoc(doc, () => `Model: ${key}`),
        bones: getBones(value),
      })
    );
  }

  const modern = imp["minecraft:geometry"];
  if (Array.isArray(modern)) {
    modern
      .filter((m): m is internal.ModelModernSpec => internal.ModelModernSpec.is(m))
      .forEach((model) => {
        out.push(
          createModel({
            id: model.description.identifier,
            location: Types.Location.create(uri, content.indexOf(model.description.identifier)),
            documentation: Documentation.getDoc(doc, () => `Model: ${model.description.identifier}`),
            bones: getBones(model),
          })
        );
      });
  }

  return out;
}

function getBones(model: Pick<internal.ModelLegacySpec | internal.ModelModernSpec, "bones">): string[] {
  const out: string[] = [];

  model.bones.forEach((bone) => {
    const name = bone.name;
    if (typeof name !== "string" || name === "") return;
    if (out.includes(bone.name)) return;

    out.push(bone.name);
  });

  return out;
}

function createModel(current: Model): Model {
  // Might be inheriting another geometry, thus split it and return
  const keys = current.id.includes(':geometry') ? current.id.split(":geometry")[0] : current.id;

  return {
    ...current,
    id: keys,
  };
}
