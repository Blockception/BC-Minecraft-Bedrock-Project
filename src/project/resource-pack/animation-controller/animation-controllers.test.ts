import { process } from "./process";

describe("Resourcepack - Animation Controller", () => {
  const data = `{
    "format_version": "1.8.0",
    "animation_controllers": {
      "controller.animation.example.base_pose": { "states": { "default": { "animations": [{ "base_pose": [] }] } } }
    }
  }`;

  test("process", () => {
    const controllers = process({ getText: () => data, uri: "example" });
    expect(controllers).toMatchSnapshot();
  });
});
