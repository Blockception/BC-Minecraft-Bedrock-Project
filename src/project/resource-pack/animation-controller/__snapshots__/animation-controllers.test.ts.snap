// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Resourcepack - Animation Controller Process 1`] = `
[
  {
    "animations": {
      "defined": Set {},
      "using": Set {
        "base_pose",
      },
    },
    "documentation": "RP Animation Controller: 'controller.animation.example.base_pose'",
    "id": "controller.animation.example.base_pose",
    "location": {
      "position": 71,
      "uri": "example",
    },
    "molang": MolangSet {
      "assigned": Set {},
      "cache": MolangSyntaxCache {
        "_data": Map {},
      },
      "functions": Set {},
      "using": Set {},
    },
    "particles": {
      "defined": Set {},
      "using": Set {},
    },
    "sounds": {
      "defined": Set {},
      "using": Set {},
    },
  },
]
`;

exports[`Resourcepack - Animation Controller process 1`] = `
[
  {
    "animations": {
      "defined": Set {},
      "using": Set {
        "base_pose",
      },
    },
    "documentation": "RP Animation Controller: 'controller.animation.example.base_pose'",
    "id": "controller.animation.example.base_pose",
    "location": {
      "position": 71,
      "uri": "example",
    },
    "molang": MolangSet {
      "assigned": Set {},
      "cache": MolangSyntaxCache {
        "_data": Map {},
      },
      "functions": Set {},
      "using": Set {},
    },
    "particles": {
      "defined": Set {},
      "using": Set {},
    },
    "sounds": {
      "defined": Set {},
      "using": Set {},
    },
  },
]
`;
